<a href="https://colab.research.google.com/github/tahereh220/Python-I-II/blob/main/class_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cells
A notebook is a list of cells. Cells contain either explanatory text or executable code and its output. Click a cell to select it.

## Code cells
Below is a **code cell**. Once the toolbar button indicates CONNECTED, click in the cell to select it and execute the contents in the following ways:

* Click the **Play icon** in the left gutter of the cell;
* Type **Cmd/Ctrl+Enter** to run the cell in place;
* Type **Shift+Enter** to run the cell and move focus to the next cell (adding one if none exists); or
* Type **Alt+Enter** to run the cell and insert a new code cell immediately below it.

There are additional options for running some or all cells in the **Runtime** menu.


In [ ]:
a = 10
a

10

## Text cells
This is a **text cell**. You can **double-click** to edit this cell. Text cells
use markdown syntax. To learn more, see our [markdown
guide](/notebooks/markdown_guide.ipynb).

You can also add math to text cells using [LaTeX](http://www.latex-project.org/)
to be rendered by [MathJax](https://www.mathjax.org). Just place the statement
within a pair of **\$** signs. For example `$\sqrt{3x-1}+(1+x)^2$` becomes
$\sqrt{3x-1}+(1+x)^2.$


##Creating a Comment
Comments starts with a #, and Python will ignore them:

In [ ]:
#This is a comment
print("Hello, World!")

Hello, World!


##Python Data Types
Variables can store data of different types.
Python has the following data types built-in by default, in these categories:

*   Text Type: **str**

>>\n : Line change.
>>
>>\t : Tab.
>>
>>\\ : Backslash character.
>>
>>r : Carriage return.
>>
>>\a : Bell.
>>
>>\b : Back space.


*   Numeric Types: **int, float, complex**
*   Boolean Type: **bool**
*   Sequence Types: **list, tuple, range**
*   Mapping Type: **dict**
*   Set Types: **set, frozenset**


In [ ]:
print("type(´Hi´) : ", type("Hi!"))

type(´Hi´) :  <class 'str'>


In [ ]:
print("type(5) = ", type(5.0))

type(5) =  <class 'float'>


In [ ]:
print("type(0o123)", type(0o123))   # Octal

type(0o123) <class 'int'>


In [ ]:
print("type(0xFFAA)", type(0xFFAA)) # Hexadecimal

type(0xFFAA) <class 'int'>


In [ ]:
print("type(5.3) : ", type(5.3))

type(5.3) :  <class 'float'>


In [ ]:
print("type(1+5j) : ", type(1+5j))

type(1+5j) :  <class 'complex'>


In [ ]:
print("type(True) : ", type(True))

type(True) :  <class 'bool'>


In [ ]:
print("type(False) : ", type(False))

type(False) :  <class 'bool'>


In [ ]:
print("type(range(2,6)) : ", type(range(2,6)), range(2,6))

type(range(2,6)) :  <class 'range'> range(2, 6)


In [ ]:
print("type([1 , 2, 2, 100.5]) : ", type([1 , 2, 2, 100.5]))

type([1 , 2, 2, 100.5]) :  <class 'list'>


In [ ]:
print("type((1 , 2, 2, 100.5)) : ", type((1 , 2, 2, 100.5)))

type((1 , 2, 2, 100.5)) :  <class 'tuple'>


In [ ]:
print("type({1 , 2, 2, 100.5}) : ", type({1 , 2, 2, 100.5}))

type({1 , 2, 2, 100.5}) :  <class 'set'>


In [ ]:
print("He said:\n'Python is very easy'")

He said:
'Python is very easy'


In [ ]:
print('He said, "Python is very easy"')

He said, "Python is very easy"


In [ ]:
print(3 + 6)

9


In [ ]:
print(6 + 7.0)

13.0


In [ ]:
print(4 * 7)

28


In [ ]:
print(49 / 7)  # division always returns float

7.0


##Python Arithmetic Operators
 \+ : Addition	
> x + y

\- : Subtraction
> x - y	

\* : Multiplication
> x * y

/ : Division
> x / y	

\% : Modulus
> x % y	

\** : Exponentiation
> x ** y	

\// : Floor division
> x // y


In [ ]:
3 ** 2

9

In [ ]:
9 ** 0.5

3.0

In [ ]:
9 ** (-1)

0.1111111111111111

In [ ]:
69 / 2

34.5

In [ ]:
69 // 2

34

In [ ]:
69 % 2

1

##Python string Concatenation (+) and Repetition (*)

In [ ]:
str1 = 'First string, '
str2 = 'Second string.'
str1 + str2

'First string, Second string.'

In [ ]:
str = 'Python '
print(str * 3)
a = 2
x = "5"
y = 5.0
#print(a+x) 
print(type(a + y))

Python Python Python 
<class 'float'>


##Python string Indexing and Slicing:

* In Python, indices start from 0. So, the first character corresponds to the index 0!
* Negative index means position from the end; index -1 corresponds to the last character. 
* An index range a:b includes a and excludes b.


In [ ]:
str = 'Life is too short, You need Python!'
print(str[5])
str[5]

i


'i'

In [ ]:
str[3]

'e'

In [ ]:
str[-1]

'!'

In [ ]:
str[4:8]

' is '

In [ ]:
str[:17]

'Life is too short'

In [ ]:
str[19:]

'You need Python!'


##Some of the useful Python string methods:

* **x.count(str)** : Count the string pattern.
* **x.find(str)** : Position of  the first match. -1 if no match is found.
* **x.index(str)** : Position of the first match. Error if no match is found.
* **len(x)** : String length. This is not a string method but a built-in Python function.
* **x.lstrip()**	Trim left white spaces.
* **x.rstrip()** : Trim right white spaces.
* **x.strip()** : Trim white spaces from both sides.
* **x.replace(str1, str2)** : Replace string from str1 to str2.
* **x.join(str_list)** : Join the string list items using x as connector.
* **x.split(y)** : Split a string by y.
* **x.upper()** : Convert to the upper case.
* **x.lower()** : Convert to the lower case.


In [ ]:
x = 'Python'
len(x)				# String length. This is a Python built-in function.


6

In [ ]:
x.count('h')				# Counts the 'h' character.

1

In [ ]:
x.upper()				# Convert to the uppercase.

'PYTHON'

In [ ]:
x.lower()				# Convert to the lowercase.

'python'

In [ ]:
x = 'Python'
x.find('o')				# Position of the 'o'.


4

In [ ]:
x.find('w')				# Look for the position of the 'w'. No match.


-1

In [ ]:
x.index('o')				# Look for the position of the 'o'.


4

In [ ]:
x.index('w')				# Look for the position of the 'w'. No match.

ValueError: ignored

In [ ]:
x = 'Life is too short, You need Python!'
print(x.split(' '))                                                                      		  # The output is a list (*).)
print(x.split("o"))

['Life', 'is', 'too', 'short,', 'You', 'need', 'Python!']
['Life is t', '', ' sh', 'rt, Y', 'u need Pyth', 'n!']


In [ ]:
y = x.split(' ')
print("y = ", y)
a = ' '
print("a.join(y) = ", a.join(y))


y =  ['Life', 'is', 'too', 'short,', 'You', 'need', 'Python!']
a.join(y) =  Life is too short, You need Python!


##Python Logical Operators
* **and** : Returns True if both statements are true	
* **or** : Returns True if one of the statements is true
* **not** : Reverse the result, returns False if the result is true


In [ ]:
False and True

False

In [ ]:
"h" and False

False

In [ ]:
not(2 == 3)

True

In [ ]:
not(2 >= 1)

False

In [ ]:
5 < 4

False

##Python Variables

* Variables are containers for storing data values.
* Python has no command for declaring a variable.
* A variable is created the moment you first assign a value to it (dynamic type).
* Name of a variable can start with a letter or _
* Variable names are case sensitive

In [ ]:
x = 5
print(type(x))

<class 'int'>


In [ ]:
y = "hello"
print(type(y))

<class 'str'>


In [ ]:
x = y
print(type(x))

<class 'str'>


In [ ]:
x = 2
y = 1.5
x = x + y
print(x)

3.5


##Variable Naming Conventions

* Snakecase: Words are delimited by an underscore.
* Pascalcase: Words are delimited by capital letters.
* Camelcase: Words are delimited by capital letters, except the initial word.

In [ ]:
my_age = 25  # snake case variable naming

myAge = 24  # camel case variable naming

MyAge = 23  # pascal case

print(MyAge == myAge)
print(MyAge)
print(myAge)

PI = 3.14  # used for constants

False
23
24


## Follow practice steps on 'ex_0101.ipynb'
## Follow practice steps on 'ex_0102.ipynb'

##Composite Data Types in Python

* **List**: stores multiple values that may or may not be of the same type.
* **Tuple**: similar to the list with some restrictions. Tuples are immutable objects (*).
* **Dictionary**: associative array.
* **Set**: unordered collection of values without repetition. 


**A list contains comma separated values within square brackets.**

In [ ]:
a = [ ]				                                         # An empty list.

In [ ]:
b = [1, 2, 3]			                                     # A list with numeric values.

In [ ]:
c = ['Life', 'is', 'too', 'short']		                 # A list with strings.

In [ ]:
d = [1, 2, 'Life', 'is']			                         # A list with mixed data type values.	

In [ ]:
d = [1, 2, 'Life', 'is']			                         # A list with mixed data type values.

In [ ]:
e = [1, 2, ['Life', 'is']]		                         # A nested list.

In [ ]:
e[1]				                                           # Indexing of a list.

2

In [ ]:
e[2]

['Life', 'is']

In [ ]:
e[2][1]

'is'

In [ ]:
a = [1, 2, 3, 4, 5]

**Slicing a list:**

In [ ]:
a[:2]                                      # From the index 0 to 1 (inclusive).

[1, 2]

In [ ]:
a[2:]                       		           # From the index 2 to the end.

[3, 4, 5]

In [ ]:
a[:]			          	                     # The whole list!

[1, 2, 3, 4, 5]

In [ ]:
a = [1, 2, 3, ['a', 'b', 'c'], 4, 5]

In [ ]:
a[2:5]

[3, ['a', 'b', 'c'], 4]

In [ ]:
a[3][:2]

['a', 'b']

**Concatenation (+) :**

To join two list objects, use the + operator.

In [ ]:
a = [1, 2, 3]
b = [4, 5, 6]
a + b

[1, 2, 3, 4, 5, 6]

**Repetition (*):**

To repeat a list object by a given number of times, use the * operator. 

In [ ]:
a = [1, 2, 3]
a * 3

[1, 2, 3, 1, 2, 3, 1, 2, 3]

**Modifying a list: change one or multiple entries :**



In [ ]:
a = [1, 2, 3]
a[1] = -1
a

[1, -1, 3]

In [ ]:
a[1:3]

[-1, 3]

In [ ]:
a[1:3] = [3, 5, 7, 9, 11]
a

[1, 3, 5, 7, 9, 11, 7, 9, 11]

**Modifying a list: careful with the following differences.**

In [ ]:
a = [1, 2, 3]
a[1:2] = [-1, -2, -3]			             # Partial change of the list values.
a

[1, -1, -2, -3, 3]

In [ ]:
a = [1, 2, 3]
a[1] = [-1, -2, -3]			                # An entry is replaced by another list.
a

[1, [-1, -2, -3], 3]

**Modifying a list: partial removal.**

In [ ]:
a = [1, 2, 3, 4, 5]
a[1:3] = []				# Partial removal.
a

[1, 4, 5]

In [ ]:
a = [1, 2, 3, 4, 5]
del a[2]				# Removes a specific entry using del keyword
a

[1, 2, 4, 5]

##Python list methods:
* **x.insert(pos, val) :** Insert a value at a specified position.
* **x.remove(val) :** Remove a specific value from a list.
* **x.pop() :** Remove and return the last value of a list.
* **x.count(val) :** Count the appearances of a value in a list.
* **x.extend(y) :** Extend the list x with another list y. Means  x = x + y.
* **x.append(val) :** Appends a value as an entry.
* **x.sort() :** Sorting of a list.
* **x.reverse() :** Reversal of the entry ordering.
* **x.index(val) :** Looks for the position of a specific value. Error if no match is found.
* **len(x) :** Total number of entries in a list. A built-in Python function.




In [ ]:
a = [3, 1, 5, 2, 4]
a.sort()	      			# Sorting of a list object (lasting effect).
a

[1, 2, 3, 4, 5]

In [ ]:
a = [1, 2, 3]
a.append(4)			    	# A value is added as an entry. 		
a.append([5,6])		 		# A list is added as an entry.
a

[1, 2, 3, 4, [5, 6]]

In [ ]:
a = [3, 1, 5, 2, 4]
a.sort(reverse=True)			# Sorting in reverse order.		
a

[5, 4, 3, 2, 1]

In [ ]:
a = [1, 2, 3, 4, 5]
a.reverse()				       # Reverse the order. Different from sorting.	
a


[5, 4, 3, 2, 1]

In [ ]:
a = [3, 1, 5, 2, 4]
sorted(a)			         	# Displays a sorted view.

[1, 2, 3, 4, 5]

In [ ]:
a                       # No permanent changes.

[3, 1, 5, 2, 4]

In [ ]:
a.index(5)

2

##Composite Data Types : Tuple (1/5)

In [ ]:
a = ( )
a				                         # An empty tuple.

()

In [ ]:
b = (1, 2, 3)
b[2]				# A tuple with numeric values.
b[2] = 4  #error

TypeError: ignored

In [ ]:
c = ('Life', 'is', 'too', 'short')		                     #A tuple with strings.
c

('Life', 'is', 'too', 'short')

In [ ]:
d = (1, 2, 'Life', 'is')			                         # A tuple with mixed data type values.	
d

(1, 2, 'Life', 'is')

In [ ]:
e = (1, 2, ('Life', 'is'))
e

(1, 2, ('Life', 'is'))

In [ ]:
e[1]

2

In [ ]:
a = (1, 2, 3, 4, 5)
print(a[:2])                                                                                                      # From the index 0 to 1 (inclusive).      
print(a[2:])                                         		                         # From the index 2 to the end.
print(a[:])				                         # The whole tuple!	



(1, 2)
(3, 4, 5)
(1, 2, 3, 4, 5)


In [ ]:
a = (1, 2, 3)
b = (4, 5, 6)
a + b


(1, 2, 3, 4, 5, 6)

In [ ]:
a = (1, 2, 3)
a[2] = 5
a = b * 3

del a[1]



TypeError: ignored

In [ ]:
a = { }					# An empty dictionary.
b = {'NAME': 'JAMES', 
     'GENDER': 'MALE', 
     'AGE': 35}	# Numbers and strings as values. 	
c = {1: 'Life', 2: 'is', 3: 'short' }		
d = {'x':[1,2,3], 'y':[4,5,6]}			# Lists are used as values.
e = {1:b, 2:c}
b['NAME']
print(e[1])
print(d['x'])
print(c[3])
print(d['x'][1])

{'NAME': 'JAMES', 'GENDER': 'MALE', 'AGE': 35}
[1, 2, 3]
short
2


In [ ]:
a = {}				                         # An empty dictionary.		
a['NAME'] = 'JAMES'			# Add a pair.
a['GENDER'] = 'MALE'			# Add a pair.
a['AGE'] = 35				# Add a pair.
a


{'AGE': 35, 'GENDER': 'MALE', 'NAME': 'JAMES'}

In [ ]:
del a['AGE'] 				# Delete a pair.
a

{'GENDER': 'MALE', 'NAME': 'JAMES'}

In [ ]:
a['NAME'] = 'JOHN' 			# Change the value of a pair. 
a

{'GENDER': 'MALE', 'NAME': 'JOHN'}

In [ ]:
a.items()				# Show the content of a dictionary. 

dict_items([('NAME', 'JOHN'), ('GENDER', 'MALE')])

In [ ]:
a.clear()
a

{}

In [ ]:
x = [1, 2, 3, 4]
print(x[2])
print(x[10])

3


IndexError: ignored

In [ ]:
a = {'name': 'JONE', 'gender': 'MALE', 'age': 35} 
print(a['wage'])

KeyError: ignored

In [ ]:
a = {'name': 'JONE', 'gender': 'MALE', 'age': 35} 
print(a.get('name'))
print(a['name'])
print(a.get('wage', 5))
print(a)

JONE
JONE
5
{'name': 'JONE', 'gender': 'MALE', 'age': 35}


In [ ]:
a = {'name': 'JOHN', 'gender': 'MALE', 'age': 35} 
print('name' in a)
print('wage' in a)

True
False


In [ ]:
a = set()				# An empty set.
b = set([1,2,3,3,3,3,4,5])
c = tuple([1,2,3,3,3,3,4,5])
print(a)
print(b)
print(c)
print(type(a))
print(type(b))
print(type(c))
print(type([1,2,3,3,3,3,4,5]))


set()
{1, 2, 3, 4, 5}
(1, 2, 3, 3, 3, 3, 4, 5)
<class 'set'>
<class 'set'>
<class 'tuple'>
<class 'list'>


In [ ]:
s1 = set([1, 2, 3, 4, 5])
s2 = set([4, 5, 6, 7, 8])
print(s1 & s2)				# Intersection of s1 and s2.	
print(s1.intersection(s2))
print(s1 | s2)
print(s1.union(s2))
print(s1-s2)
print(s1.difference(s2))
s2.add(1)
print(s2)
s2.update([7,9,11])
print(s2)
s2.remove(6)
print(s2)


{4, 5}
{4, 5}
{1, 2, 3, 4, 5, 6, 7, 8}
{1, 2, 3, 4, 5, 6, 7, 8}
{1, 2, 3}
{1, 2, 3}
{1, 4, 5, 6, 7, 8}
{1, 4, 5, 6, 7, 8, 9, 11}
{1, 4, 5, 7, 8, 9, 11}


In [ ]:
m = input()
print(m)
if x in s2:
  s2.remove(m)
  print(s2)
else:
  print(m, " is not in the set")

6
6
6  is not in the set


##Quiz 0101 

**1)**. In the class A, students' heights are:

161.5
165.9
168.7
163.1
170.8
167.2

Calculate the mean height by using Python 'list' and 'for' loop:

In [193]:
A = [161.5,165.9,168.7,163.1,170.8,167.2]
sum = 0
i = 0
for a in A:
    print(a)
    sum += a
    i = i+1
print(sum/len(A))

161.5
165.9
168.7
163.1
170.8
167.2
166.20000000000002


**2)**. In the class B, students' heights are:

172.5
162.9
161.7
160.1
161.8
169.2

Calculate the mean height by using Python 'list' and 'for' loop:

In [194]:
B = [172.5,162.9, 161.7, 160.1, 161.8, 169.2] 
sum = 0
for b in B:
    sum += b
print(sum/len(B))

164.70000000000002


**3).** The class A and the class B are joined together to form the class AB. Calculate the mean height by using Python 'list' and 'for' loop:

In [195]:
AB = A + B
sum = 0
for ab in AB:
    sum += ab
print(sum/len(AB))

165.45000000000002


**4).** Sort the heights of the class AB.

In [196]:
AB.sort()
print(AB)

[160.1, 161.5, 161.7, 161.8, 162.9, 163.1, 165.9, 167.2, 168.7, 169.2, 170.8, 172.5]


**5)**. The class C is formed with students with the height less than 165. Make a list 'C' containing the heights of the class C.

In [197]:
C =[]
for ab in AB:
    if (ab < 165):
        C.append(ab)
print(C)

[160.1, 161.5, 161.7, 161.8, 162.9, 163.1]


**6).** The class D is formed with students with the height larger than 160 and less than 170. Make a list containing the heights of the class D.

In [ ]:
D =[]
for ab in AB:
    if (ab > 160 and ab < 170):
        D.append(ab)
print(D)